README
README for csmanager


NEWS
NEWS for csmanager


gopt.c
/*      gopt.c
 *
 *  Copyright 2017 Robert L (Bob) Parker rlp1938@gmail.com
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 *  MA 02110-1301, USA.
*/

#include "files.h"
#include "dirs.h"
#include "str.h"
#include "gopt.h"


options_t process_options(int argc, char **argv)
{
	synopsis = thesynopsis();
	helptext = thehelp();
	optstring = ":hd:Dl:c:";

	/* declare and set defaults for local variables. */

	/* set up defaults for opt vars. */
	options_t opts = {0};	// assumes defaults all 0/NULL
	// initialise non-zero defaults below.


	int c;

	while(1) {
		int this_option_optind = optind ? optind : 1;
		int option_index = 0;
		static struct option long_options[] = {
		{"help",		0,	0,	'h'},
		{"dirs-from",	1,	0,	'd'},
		{"dot-files",	0,	0,	'D'},
		{"dot-files-dir",	1,	0,	'l'},
		{"cloud-target",	1,	0,	'c'},
		{0,	0,	0,	0 }
		};

		c = getopt_long(argc, argv, optstring,
                        long_options, &option_index);

		if (c == -1)
			break;

		switch (c) {
		case 0:
			switch (option_index) {
			} // switch()
		break;
		case 'h':
		dohelp(0);
		break;
		case 'd':
			opts.dirs_from = xstrdup(optarg);
			break;
		case 'D':
			opts.do_dot_files = 1;
			break;
		case 'l':
			opts.dot_files_dir = xstrdup(optarg);
			break;
		case 'c':
			opts.cloud_target = xstrdup(optarg);	// --cloud-target
			break;
		case ':':
			fprintf(stderr, "Option %s requires an argument\n",
					argv[this_option_optind]);
			dohelp(1);
		break;
		case '?':
			fprintf(stderr, "Unknown option: %s\n",
					 argv[this_option_optind]);
			dohelp(1);
		break;
		} // switch()
	} // while()
	return opts;
} // process_options()

void dohelp(int forced)
{
  if(strlen(synopsis)) fputs(synopsis, stderr);
  fputs(helptext, stderr);
  exit(forced);
} // dohelp()

char *thesynopsis(void)
{ /* Moved this text off the top of the page. */
	char *ret =
  "\tSYNOPSIS\n"
  "\t\tcsmanager [option] dirname\n\n"
  "\tDESCRIPTION\n"
  "\tcsmanager assists with managing directories to be syncronised with"
  " a\n\tnextcloud server. The server is to be set up to sync from\n\t"
  "$HOME/nextcloud/. The dirs to be synced by default, are all named "
  "dirs\n\tin $HOME that are not prefixed with '.' ie hidden dirs.\n"
  "\tThe program creates required dirs under nextcloud/ and then\n\t"
  "executes synclink to make hard links to of all files from the "
  "named\n\tdirs."
  "\n\n";
	return ret;
} // thesynopsis()

char *thehelp(void)
{
	char *ret =
  "\tOPTIONS\n"
  "\t-h, --help\n"
  "\tOutputs this help message and then quits.\n\n"
  "\t-d, --dirs-from dir_names_file\n"
  "\tSyncs a specific list of dirs rather than every non hidden "
  "dir in\n\t$HOME. This is especially useful for users taking "
  "advantage of\n\tfree sites made available to users with small "
  "storage needs.\n\n"
  "\t-D, --dot-files \n"
  "\tWhen selected this option causes the program to check the dot "
  "dirs\n\tunder $HOME and any such dirs that have files newer than "
  "when this was\n\tlast run will be tarballed and copied into the "
  "dot-files-dir.\n\n"
  "\t-l, --dot-files-dir dir_name\n"
  "\tBy default dot files tarballs are copied into $HOME/dotty. "
  "You can name\n\tanother dir for this purpose. However, this is "
  "not a persistent change.\n\tThe tarballs storing dot dirs data "
  "are eg named like this:\n\t$HOME/.config/ becomes "
  "$HOME/dotty/config.tgz and so on.\n\n"
  "\tFILES\n"
  "\tThere is a file $HOME/dottim the modification time of which is "
  "set to\n\tthe time of completion of the last dot-files run. Initially "
  "this file\n\tis dated '1970-01-01' so that all dot dirs are tarballed"
  " on the first\n\ttime the option is selected.\n"
  ;
	return ret;
} // thehelp()


gopt.h
/*      gopt.h
 *
 *  Copyright 2017 Robert L (Bob) Parker rlp1938@gmail.com
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 *  MA 02110-1301, USA.
*/

#ifndef GOPT_H
#define GOPT_H
char *optstring;
char *helptext;
char *synopsis;

typedef struct options_t {	// to be initialised with required vars.
	char	*dirs_from;		// -d, --dirs-from
	int	 	do_dot_files;	// -D, --dot-files
	char	*dot_files_dir;	// -l, --dot-files-dir
	char	*cloud_target;	// -c, --cloud-target

} options_t;

void dohelp(int forced);
options_t process_options(int argc, char **argv);
char *thehelp(void);
char *thesynopsis(void);

#endif


files.c
/*    files.c
 *
 * Copyright 2017 Robert L (Bob) Parker rlp1938@gmail.com
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA.
*/

/* The purpose of files.[h|c] is to provide utility functions for
 * manipulating file system objects other than directories. For the
 * latter see dirs.[h|c].
 * */

#include "files.h"

void
writestrarray(char **list)
{ /* output the strings to console - must be NULL terminated. */
	size_t i;
	for (i = 0; list[i] ; i++)
		printf("%s\n", list[i]);
} // writestrarray()

char
**getfile_str(const char *path)
{ /* Read file at path. File to have lines terminated with '\n'.
   * Returns list of C strings with NULL terminated list.
*/
	mdata *md = readfile(path, 1, 0);
	size_t n = memlinestostr(md);
	if (!n)
	{
		fprintf(stderr, "File %s contains no line delimeters,\n", path);
		exit(EXIT_FAILURE);
	}
	return memblocktoarray(md, 0);
} // getfile_str()

time_t
getfile_mtime(const char *path)
{/* get the modification time of a file if it exists.
  * No interest in micro seconds for this purpose.
*/
	struct stat sb;
	int res = lstat(path, &sb);
	if (res == -1) return 0;	// 1970-01-01
	return sb.st_atime;
} // getfile_mtime()

int
xsystem(const char *cmd, int fatal)
{ /* Runs system() and processes the results.
   * If fatal is non zero all non zero results from the child will be
   * fatal but there can be circumstances where the result is needed by
   * the caller.
*/
	const int status = system(cmd);
	if (status == -1) {	// this always fatal
		fprintf(stderr, "system failed to execute: %s\n", cmd);
		exit(EXIT_FAILURE);
	}
	int res;
	if (WIFEXITED(status)) {	// Child has terminated
		res = WEXITSTATUS(status);
		if (res) {
			fprintf(stderr, "Command \"%s\" returned non-zero result:"
						" %d\n" ,cmd, res);
			if (fatal) exit(EXIT_FAILURE);
		}
	}
	return res;
} // xsystem()

void
dumpstrblock(const char *tmpfn, mdata *md)
{ /* Dumps the block of C strings named by md to the file named by
   * tmpfn. The block will be altered to a block of lines terminated by
   * '\n', then written to tmpfn and restored to C strings before
   * returning. Tmpfn may be "-" to write to stdout.
*/
	memstrtolines(md);
	writefile(tmpfn, md->fro, md->to, "w");
	memlinestostr(md);
} // dumpstrblock()

ino_t
getinode(char *path)
{	/* return the inode number if the path exists, if not abort */
	struct stat sb;
	if (lstat(path, &sb) == -1) {
		perror(path);
		exit(EXIT_FAILURE);
	}
	return sb.st_ino;
} // getinode()

void
touch(const char *fn)
{/* Emulates the simplest use of the shell touch command. */
	FILE *fp = dofopen(fn, "a");	// avoid zeroing an existing file.
	dofclose(fp);
} // touch()

void
str2file(const char *fn, const char *s)
{/* Writes the C string s to file fn. The terminating '\0' of s will be
  * replaced by '\n'.
*/
	size_t len = strlen(s);
	char *buf = xmalloc(len + 1);
	strcpy(buf, s);
	buf[len] = '\n';
	writefile(fn, buf, buf+len+1, "w");
	free(buf);
} // str2file()

FILE
*dofopen(const char *fn, const char *fmode)
{	/* fopen() with error handling. */
	FILE *fpx = fopen(fn, fmode);
	if (!fpx) {
		perror(fn);
		exit(EXIT_FAILURE);
	}
	return fpx;
} // untitled()

void
dofclose(FILE *fp)
{	// fclose() and aborts on error.
	int res = fclose(fp);
	if (res == EOF) {
		perror("fclose()");
		exit(EXIT_FAILURE);
	}
} // dofclose()

void
writefile(const char *filename, char *fro, char *to, const char *fmode)
{	/* write a block of data starting at fro, up to 'to', using mode
	*  fmode. Do nothing if to - fro <= 0.
	*/
	off_t len = to - fro;
	if (len <= 0) return;
	FILE *fpo;
	int closeit = 1;
	if (strcmp("-", filename) == 0) {
		fpo = stdout;
		closeit = 0;
	}
	else fpo = dofopen(filename, fmode);
	size_t written = fwrite(fro, 1, (size_t)len, fpo);
	if (written != (size_t)len) {
		perror("writefile");
		fprintf(stderr,
				"Expected to write: %ld bytes, but wrote %lu bytes.\n",
				len, written);
		exit(EXIT_FAILURE);
	}
	if (closeit) dofclose(fpo);
} // writefile()

void
strblocktolines(char *fro, char *to)
{	/* takes a block of contiguous C strings and formats them to
	 * printable lines, ie replace '\0' with '\n'.
	*/
	char *cp = fro;
	while (cp < to) {
		if (*cp == 0) *cp = '\n';
		cp++;
	}
} // strblocktolines()


mdata
*readfile(const char *path, int fatal, size_t extra)
{	/* If fatal is 0 will not terminate with error if path does not
	 * exist, but will return NULL instead. All other errors are always
	 * fatal. If extra is non-zero will provide extra space, init to 0.
	*/
	mdata  *ret = NULL;
	if (exists_file(path)) {
		ret = malloc(sizeof(mdata ));
		if (!ret) {
			fputs("Out of memory.\n", stderr);
			exit(EXIT_FAILURE);
		}
		size_t fsize = getfsize(path);
		size_t blocksize = fsize + extra;
		ret->fro = malloc(blocksize);
		memset(ret->fro, 0, blocksize);
		FILE *fp = dofopen(path, "r");
		size_t bread = fread(ret->fro, 1, fsize, fp);
		dofclose(fp);
		if (bread != fsize) {
			fprintf(stderr,
			"Expected to get %lu bytes, but got %lu bytes.\n",
			fsize, bread);
			exit(EXIT_FAILURE);
		}
		ret->to = ret->fro + fsize;
		ret->limit = ret->fro + blocksize;
	} else if (fatal) {
		perror(path);
		exit(EXIT_FAILURE);
	}
	return ret;
} //readfile()

int
exists_file(const char *path)
{	/* returns 1 if I can stat the file, 0 otherwise */
	struct stat sb;
	int res = stat(path, &sb);
	if (res == -1) {
		res = 0;
	} else {
		res = 1;
	}
	if (res && S_ISREG(sb.st_mode)) {
		res = 1;
	} else {
		res = 0;
	}
	return res;
} // exists_file()

off_t
getfsize(const char *path)
{	/* returns file size if path exists, fatal otherwise */
	struct stat sb;
	int res = stat(path, &sb);
	if (res == -1) {
		perror(path);
		exit(EXIT_FAILURE);
	}
	return sb.st_size;
} // getfsize()

mdata
*getconfigfile(char *pname, char *cfgfile)
{/* Read the content of $HOME/.config/pname/cfgfile */
	char path[PATH_MAX];
	sprintf(path, "%s/.config/%s/%s", getenv("HOME"), pname, cfgfile);
	mdata *cfd = readfile(path, 1, 0);
	return cfd;
} // getconfigfile()

void
copyfile(const char *pathfro, const char *pathto)
{/* Does a file copy in user space. */
	mdata *fd = readfile(pathfro, 1, 0);
	writefile(pathto, fd->fro, fd->to, "w");
} // copyfile()

void
dolink(const char *fr, const char *to)
{/* link() with error handling. */
	if (link(fr, to) == -1) {
		perror(to);
		perror(fr);	// don't know what caused the snafu
		exit(EXIT_FAILURE);
	}
} // dolink()

char
*cfg_getparameter(const char *prn, const char *fn, const char *param)
{ /* Return the string that param points to. */
	char *pp = cfg_pathtofile(prn, fn);
	mdata *md = readfile(pp, 1, 0);
	free(pp);
	memlinestostr(md);
	char *p = memmem(md->fro, md->to - md->fro, param, strlen(param));
	if (!p) {
		fprintf(stderr, "No such parameter: %s\n", param);
		free_mdata(md);
		exit(EXIT_FAILURE);
	}
	char *eq = strchr(p, '=');
	if (!eq) {
		fprintf(stderr, "Malformed parameter line: %s\n", p);
		free_mdata(md);
		exit(EXIT_FAILURE);
	}
	eq++;	// get past '='
	trimspace(eq);
	char *ret = xstrdup(eq);
	free_mdata(md);
	return ret;
} // cfg_getparameter()


csmanager.c
/*    csmanager.c
 *
 * Copyright 2017 Robert L (Bob) Parker rlp1938@gmail.com
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA.
**/

#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <dirent.h>
#include <fcntl.h>
#include <stdarg.h>
#include <getopt.h>
#include <ctype.h>
#include <limits.h>
#include <linux/limits.h>
#include <libgen.h>
#include <errno.h>
// typdefs/structs here.
typedef struct ops_t {
	char *dirname;		// the name of a dir or a file listing dirs.
	char *cloud_target;	// eg Dropbox or Nextcloud;
	int do_master_dir;	// a dir, not a file listing dirs.
	int do_dot_dirs;	// 0 is avoid dot dirs, 1 is dot dirs only.
	time_t dd_updated;	// last time we processed dot dirs.
	char *dotfiles_dir;	// the dir to send dot files tarballs to.
	char *timestamp;	// path to the timestamp for this process.
	char **files_list;	// list of dirs to process.
	char **excludes;	// list of dirs to exclude eg $HOME/Dropbox etc.
} ops_t;
#include "str.h"
#include "dirs.h"
#include "files.h"
#include "gopt.h"

static int
nc_recursedir(ops_t *operation, const char *dirname);
static char
*timestamp_name(void);
static void
process_dotdirs(ops_t *operation);
static void
check_args(ops_t *ops, char **argv, int optind, options_t *opts);
static char
**make_dirslist(ops_t *operation);
static char
**gen_dirslist(const char *dirname, int dotsornot, char **excl_list);
static char
**gen_fileslist(const char *dirname, char **excl_list);
static char
**gen_listfromfile(const char *filename);
static char
**excl_list(const char *prname);
static void
proc_normaldirs(ops_t *operation);
static char
*target_from_option(char *tname);
static char
*target_from_file(void);


int main(int argc, char **argv)
{
	options_t opts = process_options(argc, argv);	// options
	ops_t operation = {0};	// preparation
	check_args(&operation, argv, optind, &opts);
	operation.timestamp = timestamp_name();
	operation.excludes = excl_list("csmanager");
	operation.files_list = make_dirslist(&operation); // data gathering.
	if (operation.do_dot_dirs) {	// processing
		process_dotdirs(&operation);
	} else { /* possibly replace a sub-dir name with full path. */
		if (opts.cloud_target) operation.cloud_target =
								target_from_option(opts.cloud_target);
		else operation.cloud_target = target_from_file();
		proc_normaldirs(&operation);
	}
	// and now deal with sundry files in $HOME or chosen dir.
	gen_fileslist(operation.dirname, NULL);
	return 0;
}//main()

void check_args(ops_t *operation, char **argv, int optind,
			options_t *opts)
{ /* Check presence of and validity of non-option args */
	char buf[PATH_MAX] = {0};
	if (opts->dirs_from && argv[optind]) {
		sprintf(buf, "List file: %s and master dir %s not allowed\n",
				opts->dirs_from, argv[optind]);
	} else if (!argv[optind]) {
		strcpy(buf, "No directory to process provided.\n");
	} else if (!exists_dir(argv[optind])) {
		sprintf(buf, "%s does not exist or is not a dir.\n",
				argv[optind]);
	} else if (opts->dirs_from && opts->do_dot_files) {
		sprintf(buf, "May not process dot dirs in list of dirs: %s\n",
				opts->dirs_from);
	} else if (opts->dirs_from && !exists_file(opts->dirs_from)) {
				sprintf(buf, "%s does not exist or is not a file.\n",
						opts->dirs_from);
	}

	if (strlen(buf)) {
		fprintf(stderr, "%s", buf);
		exit(EXIT_FAILURE);
	}
	// Can complete ops_t struct
	if (argv[optind]) {
		operation->do_master_dir = 1;
		operation->dirname = realpath(argv[optind], NULL);
		if (!operation->dirname) {
			perror(argv[optind]);
			exit(EXIT_FAILURE);
		}
	} else if (opts->dirs_from) {
		operation->dirname = xstrdup(opts->dirs_from);
	}
	operation->do_dot_dirs = opts->do_dot_files;
	if (opts->dot_files_dir) {
		// dot file location is initialised needed or not.
		operation->dotfiles_dir = realpath(opts->dot_files_dir, NULL);
		if (!operation->dotfiles_dir) {
			perror(argv[optind]);
			exit(EXIT_FAILURE);
		}
	} else { // dot dir tarballs go to standard place.
		sprintf(buf, "%s/dotty", getenv("HOME"));
		operation->dotfiles_dir = xstrdup(buf);
	}
	// set up the cloud target name
	if (opts->cloud_target)
	{ /* if user has optioned a target it won't be NULL */
		operation->cloud_target = target_from_option(opts->cloud_target);
	} else
	{ /* no optional target so get it from file. */
		operation->cloud_target = target_from_file();
	}
} // check_args()

char
**make_dirslist(ops_t *operation)
{/* return a list of dirs using method depending on type of data given.
**/
	if (operation->do_master_dir) {
		return gen_dirslist(operation->dirname, operation->do_dot_dirs,
							operation->excludes);
	} else {
		return gen_listfromfile(operation->dirname);
	}
} // make_dirslist()

char
**gen_dirslist(const char *dirname, int dotsornot, char **excl_list)
{/* get the dir names under dirname selecting or avoiding dot dirs,
  * then turn the data into an array of C strings.
*/
	mdata *md = init_mdata();
	const size_t meminc = 1024 * 1024;	// This instance 1 meg good.
	DIR *thedir = dopendir(dirname);
	struct dirent *de;
	while ((de = readdir(thedir))) {
		if(strcmp("..", de->d_name) == 0) continue;
		if(strcmp(".", de->d_name) == 0) continue;
		if(de->d_type != DT_DIR) continue;
		if (dotsornot) {
			if (de->d_name[0] != '.') continue;
		} else {
			if (de->d_name[0] == '.') continue;
		}
		char buf[PATH_MAX];
		strcpy(buf, dirname);
		strjoin(buf, '/', de->d_name, PATH_MAX);
		if(inlist(buf, excl_list)) continue;
		meminsert(buf, md, meminc);
	}
	doclosedir(thedir);
	char **result = memblocktoarray(md);
	free_mdata(md);
	return result;
} // gen_dirslist()

char
**gen_listfromfile(const char *filename)
{/* get the dir names contained in the named file,
  * then turn the data into an array of C strings.
*/
	mdata *md = readfile(filename, 1, 0);	// terminated with '\n'
	memlinestostr(md);	// C strings
	char **result = memblocktoarray(md);
	free_mdata(md);
	return result;
} // gen_listfromfile()

void
process_dotdirs(ops_t *operation)
{/* set up to recurse dot dirs */
	if (!exists_file(operation->timestamp)) {
		char command[PATH_MAX];
		sprintf(command, "touch -d 1970-01-01 %s",
					operation->timestamp);
		(void)xsystem(command, 1);
	}
	operation->dd_updated = getfile_mtime(operation->timestamp);
	if (!exists_dir(operation->dotfiles_dir))
		newdir(operation->dotfiles_dir, 0);
	size_t hl = home_len();
	size_t i;
	for (i = 0; operation->files_list[i] ; i++) {
		// ->files_list is NULL terminated.
		int res = nc_recursedir(operation, operation->files_list[i]);
		if (res) {
			char toname[NAME_MAX];
			strcpy(toname, operation->dotfiles_dir);	// tarballs dir.
			strjoin(toname, '/', operation->files_list[i] + hl+2,
						NAME_MAX);
			strjoin(toname, '.', "tgz", NAME_MAX);
			/* tarball name is the name of the dot dir less '.'. */
			char comm[PATH_MAX];
			sprintf(comm, "tar -czf %s %s", toname,
						operation->files_list[i]);
			(void)xsystem(comm, 0);
		}
	}
	char command[PATH_MAX];	// time stamp update
	sprintf(command, "touch %s", operation->timestamp);
	(void)xsystem(command, 1);
} // process_dotdirs()

char
*timestamp_name(void)
{
	char buf[PATH_MAX];
	strcpy(buf, getenv("HOME"));
	strjoin(buf, '/', "dottim", PATH_MAX);
	return xstrdup(buf);
} // timestamp_name()

int
nc_recursedir(ops_t *operation, const char *dirname)
{ /* Recurse the given dirname checking dir and file objects for
   * age against the time stamp file. Return 1 if a newer object is
   * found, 0 otherwise.
*/
	time_t stamp = getfile_mtime(dirname);
	if (stamp > operation->dd_updated) return 1;
	DIR *dp = dopendir(dirname);
	struct dirent *de;
	while ((de = readdir(dp))) {
		if (strcmp(de->d_name, ".") == 0 ) continue;
		if (strcmp(de->d_name, "..") == 0) continue;
		if (!(de->d_type == DT_REG || de->d_type == DT_DIR)) continue;
		char joinbuf[PATH_MAX];
		strcpy(joinbuf, dirname);
		strjoin(joinbuf, '/', de->d_name, PATH_MAX);
		stamp = getfile_mtime(joinbuf);
		if (stamp > operation->dd_updated){
			doclosedir(dp);
			return 1;
		}
		if (de->d_type == DT_DIR) {
			int result = nc_recursedir(operation, joinbuf);
			if (result) {
				doclosedir(dp);
				return 1;
			}
		}
	} // while()
	doclosedir(dp);
	return 0;
} // nc_recursedir()

char
**excl_list(const char *prname)
{/* Return list of dirs to exclude from processing. If the excludes file
  * does not exist, create it with some reasonable default values.
*/
	char fpath[PATH_MAX] = {0};
	sprintf(fpath, "%s/.config/%s/excl.lst", getenv("HOME"), prname);
	if (!exists_file(fpath))
	{	// create it
		char *cp = strstr(fpath, "excl.lst");
		*cp = 0;	// does the dir exist? If not create it.
		if (!exists_dir(fpath)) newdir(fpath, 0);
		*cp = 'e';	// restore the filename to fpath.
		FILE *fpo = dofopen(fpath, "w");
		fprintf(fpo, "%s/%s\n", getenv("HOME"), "Dropbox");
		fprintf(fpo, "%s/%s\n", getenv("HOME"), "Nextcloud");
		dofclose(fpo);
		sync();
	}
	return getfile_str(fpath);
} // excl_list()

static void
proc_normaldirs(ops_t *operation)
{/* We have an array of C strings of the form $HOME/dirname. Required to
  * hard link all the files and sub dirs under $HOME/dirname into
  * $HOME/cloud_target/dirname, creating dirs in target as needed.
*/
	size_t hl = home_len();
	size_t i;
	for (i = 0; operation->files_list[i]; i++)
	{	// it's a NULL terminated list
		char *frdir = operation->files_list[i];
		char todir[PATH_MAX];
		strcpy(todir, operation->cloud_target);
		strjoin(todir, '/', &frdir[hl+1], PATH_MAX);
		if(!exists_dir(todir)) newdir(todir, 1);
		char command[PATH_MAX];
		sprintf(command, "synclink '%s' '%s'", frdir, todir);
		int res = xsystem(command, 0);
		if (res)
		{
			fprintf(stderr, "Non zero system result: %d\n", res);
		}
	}
} // proc_normaldirs()

char
*target_from_option(char *tname)
{	/* set up the cloud directory name from tname */
	char joinbuf[PATH_MAX];
	char *cp = get_home();
	strcpy(joinbuf, cp);
	free(cp);
	strjoin(joinbuf, '/', tname, PATH_MAX);
	return xstrdup(joinbuf);
} // target_from_option()

char
*target_from_file(void)
{	/* set up the cloud directory name from config file */
	char *cfgn = getcfgfn("csmanager", "target");
	char *tmp;
	if (!exists_file(cfgn))
	{
		str2file(cfgn, "Nextcloud");
		tmp = xstrdup("Nextcloud");
	} else
	{
		mdata *md = readfile(cfgn, 0, 1);
		memlinestostr(md);
		tmp = xstrdup(md->fro);
		free_mdata(md);
	}
	free(cfgn);
	char buf[PATH_MAX];
	sprintf(buf, "%s/%s", getenv("HOME"), tmp);
	free(tmp);
	return xstrdup(buf);
} // target_from_file()

char
**gen_fileslist(const char *dirname, char **excl_list)
{/* get the file names under dirname, then turn the data into an array
  * of C strings.
*/
	mdata *md = init_mdata();
	const size_t meminc = PATH_MAX;	// This instance 4k meg good.
	DIR *thedir = dopendir(dirname);
	struct dirent *de;
	while ((de = readdir(thedir))) {
		if(strcmp("..", de->d_name) == 0) continue;
		if(strcmp(".", de->d_name) == 0) continue;
		if(de->d_type != DT_REG) continue;
		char buf[PATH_MAX];
		strcpy(buf, dirname);
		strjoin(buf, '/', de->d_name, PATH_MAX);
		if(inlist(buf, excl_list)) continue;
		meminsert(buf, md, meminc);
	}
	doclosedir(thedir);
	char **result = memblocktoarray(md);
	free_mdata(md);
	return result;
} // gen_fileslist()
